src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,227 +9,316 @@ internal struct FileStatus
+        // Checks if the main path (without following symbolic links) has the read-only attribute
+        // Only call if Refresh has been successfully called at least once
+        private bool HasReadOnlyFlag
         {
-            EnsureStatInitialized(path, continueOnError);
+            get
+            {
 #if TARGET_BROWSER
-            const Interop.Sys.Permissions readBit = Interop.Sys.Permissions.S_IRUSR;
-            const Interop.Sys.Permissions writeBit = Interop.Sys.Permissions.S_IWUSR;
+                const Interop.Sys.Permissions readBit = Interop.Sys.Permissions.S_IRUSR;
+                const Interop.Sys.Permissions writeBit = Interop.Sys.Permissions.S_IWUSR;
 #else
-            Interop.Sys.Permissions readBit, writeBit;
+                Interop.Sys.Permissions readBit, writeBit;
 
-            if (_fileStatus.Uid == Interop.Sys.GetEUid())
-            {
-                // User effectively owns the file
-                readBit = Interop.Sys.Permissions.S_IRUSR;
-                writeBit = Interop.Sys.Permissions.S_IWUSR;
+                if (_mainCache.Uid == Interop.Sys.GetEUid())
+                {
+                    // User effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRUSR;
+                    writeBit = Interop.Sys.Permissions.S_IWUSR;
+                }
+                else if (_mainCache.Gid == Interop.Sys.GetEGid())
+                {
+                    // User belongs to a group that effectively owns the file
+                    readBit = Interop.Sys.Permissions.S_IRGRP;
+                    writeBit = Interop.Sys.Permissions.S_IWGRP;
+                }
+                else
+                {
+                    // Others permissions
+                    readBit = Interop.Sys.Permissions.S_IROTH;
+                    writeBit = Interop.Sys.Permissions.S_IWOTH;
+                }
+    #endif
+                return ((_mainCache.Mode & (int)readBit) != 0 && // has read permission
+                    (_mainCache.Mode & (int)writeBit) == 0);     // but not write permission
             }
-            else if (_fileStatus.Gid == Interop.Sys.GetEGid())
+        }
+
+        internal bool HasSecondaryDirectoryFlag => HasDirectoryFlag(_secondaryCache);
+
+        // Checks if the main path is a symbolic link
+        // Only call if Refresh has been successfully called at least once
+        internal bool HasSymbolicLinkFlag =>
+            (_mainCache.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
+
+        // We track intent of creation to know whether or not we want to (1) create a
+        // DirectoryInfo around this status struct or (2) actually are part of a DirectoryInfo.
+        internal bool InitiallyDirectory { get; set; }
+
+        internal bool IsMainCacheValid => _initializedMainCache == 0;
+
+        internal bool IsSecondaryCacheValid => _initializedSecondaryCache == 0;
+
+        private bool IsValid =>
```

Comment:
nit: When I see `FileStatus.IsValid` I think of the file being valid, not file status cache. Would `HasValidCache` be a better name? 

```suggestion
        private bool HasValidCache =>
```

Reply:
Sure. I'll use `HasValidCaches` (plural).